manifold/src/manifold.cpp

```diff
@@ -102,29 +115,29 @@ Manifold::Manifold(const Mesh& mesh,
                    const std::vector<glm::ivec3>& triProperties,
                    const std::vector<float>& properties,
                    const std::vector<float>& propertyTolerance)
-    : pImpl_{std::make_unique<Impl>(mesh, triProperties, properties,
-                                    propertyTolerance)} {}
+    : pNode_(std::make_shared<CsgLeafNode>(std::make_shared<Impl>(mesh, triProperties, properties,
+                                    propertyTolerance))) {}
 
 /**
  * This returns a Mesh of simple vectors of vertices and triangles suitable for
  * saving or other operations outside of the context of this library.
  */
 Mesh Manifold::GetMesh() const {
-  pImpl_->ApplyTransform();
+  const Impl& impl = *GetCsgLeafNode().GetImpl();
```

Comment:
Actually it depends on the tree. For example, consider `A + (B + C + D)` and `A' + (B + C + D)`, in this case `B + C + D` will be evaluated 2 times because it will try to use compose to perform `A + B + C + D` and `A' + B + C + D`. Whether this is faster or combining `B + C + D -> TMP` and do `A + TMP` and `A' + TMP` is faster depends on the bounding box and the size of the manifolds.

I think it should be possible to tune this to make it combine the operands when the number of objects or the complexity (measured by number of vertices for example) exceeded a certain threshold. But this would be complicated, depends on the test case and build configuration. If users are really concerned about this, they can use some methods that depends on the `Impl` object to force the CSG tree to evaluate and cache the result.

Reply:
Would that be `GetMesh`? It kind of reminds me of OpenSCAD's `render` method, which really meant cache here. I wonder if an explicit method like that would be helpful? No rush, but something to ponder.